smack-extensions/src/main/java/org/jivesoftware/smackx/vcardtemp/provider/VCardProvider.java

```diff
@@ -18,339 +18,335 @@
-                        break;
-                    default:
+            while (!done) {
+                int eventType = parser.next();
+
+                if (eventType == XmlPullParser.START_TAG) {
+
+                    lastEventWasStart = true;
+
+                    name = parser.getName();
+
+                    if (name.equals("N")) {
+                        parseName(parser, vCard);
+                    }
+                    else if (name.equals("NICKNAME")) {
+                        vCard.setNickName(parser.nextText());
+                    }
+                    else if (name.equals("FN")) {
+                        vCard.setField("FN", parser.nextText());
+                    }
+                    else if (name.equals("URL")) {
+                        vCard.setField("URL", parser.nextText());
+                    }
+                    else if (name.equals("BDAY")) {
+                        vCard.setField("BDAY", parser.nextText());
+                    }
+                    else if (name.equals("ORG")) {
+                        parseOrg(parser, vCard);
+                    }
+                    else if (name.equals("TITLE")) {
+                        vCard.setField("TITLE", parser.nextText());
+                    }
+                    else if (name.equals("ROLE")) {
+                        vCard.setField("ROLE", parser.nextText());
+                    }
+                    else if (name.equals("TEL")) {
+                        parseTel(parser, vCard);
+                    }
+                    else if (name.equals("ADR")) {
+                        parseAddress(parser, vCard);
+                    }
+                    else if (name.equals("EMAIL")) {
+                        parseEmail(parser, vCard);
+                    }
+                    else if (name.equals("JABBERID")) {
+                        vCard.setJabberId(parser.nextText());
+                    }
+                    else if (name.equals("DESC")) {
+                        vCard.setField("DESC", parser.nextText());
+                    }
+                    else if (name.equals("PHOTO")) {
+                        parsePhoto(parser, vCard);
+                    }
+                    else {
+                        text = parser.nextText();
+                    }
+
+                    depth++;
                 }
+                else if (eventType == XmlPullParser.END_TAG) {
+
+                    depth--;
 
-                if (event == XmlPullParser.END_TAG && "vCard".equals(parser.getName())) break;
+                    if (lastEventWasStart && 0 == depth && name.length() != 0 && text.length() != 0) {
+
+                        vCard.setField(name, text);
+                    }
 
-                event = parser.next();
+                    name = "";
+                    text = "";
+                    lastEventWasStart = false;
+
+                    if (parser.getName().equals("vCard")) {
```

Comment:
Abort condition should only consider the current parser depth (see providers.md)